### PR TITLE
[ENHANCEMENT] Improve tracking for in-code data contexts

### PR DESCRIPTION
**Pull request by anthonyburdi**
### This PR is to merge #2030 and #2046 into develop, along with the following changes:

- Suppress the creation of a `store_backend_id` for inactive stores. This way when using the CLI (e.g. `great_expectations store list`), we do not attempt to connect to inactive stores. A warning message still appears for unreachable **active** stores.
- Enhance the warning message for **active** invalid store configuration to add the store_backend class name and store_name. E.g. `Invalid store configuration: store_backend_id cannot be retrieved or set. Please check the configuration of your TupleS3StoreBackend named my_s3_store_backend`. 
- Fix an oversight where `_init_stores` was called twice in `DataContext.__init__()`
- Merged in 0.13
- Change to in-code instead of in-memory and other naming changes

### From:  [ENHANCEMENT] Add a store_id property to StoreBackend #2030: 
- Add a store_backend_id property to StoreBackend & Store metaclasses
- Create and persist this store_backend_id upon instantiation of a StoreBackend
- Skip validation rules for the key of a store_backend_id, do not add suffixes (store_backend_id is saved on file/blob stores as a `.ge_store_backend_id` file)
- Test in all backends (GCS testing could be better)
- Helper methods for testing

###  From: [ENHANCEMENT] Use an existing expectation_store.store_backend_id to initialize an in-memory DataContext #2046: 
- add `_construct_data_context_id` method to choose the id of the currently-configured expectations store, if available. If not, choose the id stored in `DataContextConfig`. Keep existing functionality intact e.g. setting data_context_id via env var.
- manually init `expectations_store.store_backend_id` with the given data_context_id if there is a data_context_id in DataContextConfig and no expectation store backend is configured already
- tests for the above

Previous Design Review notes:
- See related [design doc for in-memory configurations](https://github.com/superconductive/design/blob/main/docs/20201020_in_memory_ge.md)

Design Reviewer: @jcampbell 
Code Reviewer: @talagluck

### CODE REVIEW

File: great_expectations/data_context/store/expectations_store.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+from collections import OrderedDict
```
Comment (talagluck) [inline great_expectations/data_context/store/expectations_store.py]: Is OrderedDict used here?

File: great_expectations/data_context/store/tuple_store_backend.py [hunk 2]
```diff
@@ -86,6 +92,14 @@ def _convert_key_to_filepath(self, key):
...
         # and then replaces that with a platform-specific separator if requested (the default)
         self._validate_key(key)
+        # Handle store_backend_id separately
+        if key == self.STORE_BACKEND_ID_KEY:
+            filepath = f"{self.filepath_prefix or ''}{'/' if self.filepath_prefix else ''}{key[0]}"
+            return (
+                filepath
+                if not self.platform_specific_separator
+                else os.path.normpath(filepath)
+            )
```
Comment (talagluck) [inline great_expectations/data_context/store/tuple_store_backend.py:104]: Very cool!

**Review by talagluck — APPROVED** [review]
LGTM! Just one question about an import.

File: great_expectations/data_context/store/expectations_store.py (hunk 1 above)
Comment (anthonyburdi) [reply great_expectations/data_context/store/expectations_store.py]: Removed, thank you!

**Comment by anthonyburdi** [comment]
Note: @jcampbell approved the design of these changes in a review on Mon Nov 30 2020

File: great_expectations/data_context/store/store_backend.py [hunk 3]
```diff
@@ -22,33 +22,56 @@ class StoreBackend(metaclass=ABCMeta):
...
+        Args:
+            suppress_warning: boolean flag for whether warnings are logged
+
         Returns:
             store_backend_id which is a UUID(version=4)
         """
+        if self._suppress_store_backend_id:
+            if not suppress_warning:
+                logger.warning(
+                    f"You are attempting to access store_backend_id of a store or store_backend named {self.store_name} that has been explicitly suppressed."
```
Comment (talagluck) [inline great_expectations/data_context/store/store_backend.py]:
```suggestion
                    f"You are attempting to access the store_backend_id of a store or store_backend named {self.store_name} that has been explicitly suppressed."
```

File: tests/data_context/test_data_context_in_code_config.py [hunk 4]
```diff
@@ -453,3 +453,135 @@ def test_suppress_store_backend_id_is_true_for_inactive_stores():
         ).store_backend._suppress_store_backend_id
         is False
     )
+
+
+@mock_s3
+def test_inaccessible_active_bucket_warning_messages(caplog):
+    """
+    What does this test and why?
```
Comment (talagluck) [inline tests/data_context/test_data_context_in_code_config.py]:
```suggestion
    What does this test do and why?
```

File: tests/data_context/test_data_context_in_code_config.py [hunk 5]
```diff
@@ -453,3 +453,135 @@ def test_suppress_store_backend_id_is_true_for_inactive_stores():
...
+    assert (
+        caplog.records[1].message
+        == "Invalid store configuration: Please check the configuration of your TupleS3StoreBackend named validations_S3_store"
+    )
+
+
+@mock_s3
+def test_inaccessible_inactive_bucket_no_warning_messages(caplog):
+    """
+    What does this test and why?
```
Comment (talagluck) [inline tests/data_context/test_data_context_in_code_config.py]:
```suggestion
    What does this test do and why?
```

File: tests/test_utils.py [hunk 6]
```diff
@@ -1337,6 +1338,65 @@ def safe_remove(path):
...
+    for file_name in file_name_list:
+        file_path = os.path.join(directory, file_name)
+        with open(file_path, "w") as f_:
+            f_.write(file_content_fn())
+
+
+def create_fake_data_frame():
+    return pd.DataFrame({"x": range(10), "y": list("ABCDEFGHIJ"),})
+
+
```
Comment (talagluck) [inline tests/test_utils.py]: Should these functions be in this PR?

**Review by talagluck — APPROVED** [review]
LGTM! A few minor word suggestions and one question.

File: tests/test_utils.py (hunk 6 above)
Comment (anthonyburdi) [reply tests/test_utils.py]: Nope! These are now gone after a merge from develop

**Pull request merged**